### ISSUE TITLE
Add a synchronized JointMultiAgentTrainer for cross-agent loss terms

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#5]
## Motivation

`multi_agent::{PolicyLearner, GameSimulator, Population}` implements **per-agent learner threads** that communicate via channels. This is great for league play / population training where agents update independently from a shared rollout pool.

It is **wrong** for setups that need a single loss term computed across multiple agents on the *same* minibatch, because the agents' updates are no longer synchronous and there is no batched access to all encoder outputs simultaneously.

Concrete example (Slepian-Wolf MARL P3):

```
L_red   = lambda * sum_{i<j} f(Z_i(obs), Z_j(obs))
L_total = sum_i L_ppo[i](obs, a_i, ...) + L_red
```

`L_red` requires all `N` policies' encoder outputs on the **same** batch in the same process at the same step. The current per-thread architecture doesn't support this without heavy synchronization that defeats the point.

## Current State (post-merge of `slepian-wolf-marl-port`)

After commit `40ec676 examples: joint multi-agent PPO for Slepian-Wolf P3 in Rust`, the pattern exists end-to-end as a **working example** but **not** as a reusable library trainer:

- `examples/games/bucket_brigade/train_p3.rs` (~519 lines) implements the full architecture inline: synchronized `Rollout` struct, per-agent GAE, joint PPO update with cross-agent redundancy penalty, single backward over the summed loss, N independent optimizers stepping from the shared graph.
- The reusable building blocks already landed in `src/`:
  - `src/policy/mlp.rs::MlpPolicy::encoder_features` (issue #1)
  - `src/train/ppo/trainer.rs::PPOTrainer::train_step_with_aux` (issue #2) - single-agent aux hook
  - `src/policy/multi_discrete_mlp.rs::MultiDiscreteMlpPolicy` (issue #3)
  - `src/env/games/bucket_brigade.rs::BucketBrigadeMaEnv` (issue #6)
- `src/multi_agent/mod.rs` still exposes only the per-thread `{PolicyLearner, GameSimulator, Population}` API; no `JointMultiAgentTrainer` exists in the library.

**Verdict**: keep the issue open and re-scope it as "promote the example pattern into a reusable library trainer". The example serves as the reference implementation; the work is library extraction + generalization + tests.

## Proposal

Add `src/multi_agent/joint.rs` (re-exported from `multi_agent::mod`) containing a `JointMultiAgentTrainer` that encapsulates the pattern currently inlined in `train_p3.rs`.

### Target API sketch

```rust
// src/multi_agent/joint.rs

pub struct JointTrainerConfig {
    pub num_agents: usize,
    pub rollout_steps: usize,
    pub gamma: f64,
    pub gae_lambda: f64,
    pub clip_range: f64,
    pub clip_range_vf: f64,
    pub vf_coef: f64,
    pub ent_coef: f64,
    pub n_epochs: usize,
    pub minibatch_size: usize,
    pub max_grad_norm: f64,
    pub normalize_advantages: bool,
}

/// Synchronized rollout buffer: every agent sees the same observation on every
/// step; per-agent actions / log_probs / values / rewards are stored in parallel.
pub struct JointRollout {
    pub observations: Tensor,        // [T, obs_dim]      -- shared
    pub actions: Vec<Tensor>,        // [N][T, act_shape]
    pub log_probs: Vec<Tensor>,      // [N][T]
    pub values: Vec<Tensor>,         // [N][T]
    pub rewards: Vec<Tensor>,        // [N][T]
    pub dones: Tensor,               // [T]               -- shared
}

pub struct JointStats {
    pub policy_loss: Vec<f64>,     // per agent
    pub value_loss:  Vec<f64>,
    pub entropy:     Vec<f64>,
    pub aux_loss:    f64,
    pub total_loss:  f64,
    pub clip_fraction: Vec<f64>,
    pub approx_kl:     Vec<f64>,
    pub explained_var: Vec<f64>,
}

pub struct JointMultiAgentTrainer<P: JointPolicy> {
    pub policies:  Vec<P>,
    pub optimizers: Vec<tch::nn::Optimizer>,
    pub config: JointTrainerConfig,
}

impl<P: JointPolicy> JointMultiAgentTrainer<P> {
    pub fn new(policies: Vec<P>, lr: f64, config: JointTrainerConfig) -> Result<Self> { ... }

    /// Drive a `MultiAgentEnvironment` for `config.rollout_steps` and return the
    /// synchronized rollout buffer.
    pub fn collect_rollout<E: MultiAgentEnvironment>(
        &self,
        env: &mut E,
        last_obs: &mut Vec<f32>,
    ) -> JointRollout;

    /// Joint PPO update. `aux_fn` receives the per-agent encoder-feature tensors
    /// for the current minibatch and returns an optional pre-scaled scalar loss
    /// (e.g. the cross-agent redundancy penalty). One `.backward()` flows
    /// through every encoder.
    pub fn update<F>(&mut self, rollout: &JointRollout, aux_fn: F) -> Result<JointStats>
    where
        F: FnMut(&[&Tensor /* [N] encoder features, minibatch */]) -> Option<Tensor>;
}

/// What a policy must support to participate in the joint trainer. The
/// blanket impls cover `MlpPolicy` (single-discrete) and `MultiDiscreteMlpPolicy`.
pub trait JointPolicy {
    fn get_action(&self, obs: &Tensor) -> (Tensor /*action*/, Tensor /*log_prob*/, Tensor /*value*/);
    fn evaluate_actions(&self, obs: &Tensor, actions: &Tensor) -> (Tensor, Tensor, Tensor);
    fn encoder_features(&self, obs: &Tensor) -> Tensor;
    fn device(&self) -> tch::Device;
    fn var_store(&self) -> &tch::nn::VarStore;
}
```

Mirrors the Python prototype in `bucket_brigade.training.joint_trainer`. Single process, N policies, one rollout buffer, one `total_loss.backward()` so the aux term's gradient hits every encoder.

### Refactor target

After the library lands, `examples/games/bucket_brigade/train_p3.rs` should shrink to roughly:

1. parse args + build `BucketBrigadeMaEnv` + N `MultiDiscreteMlpPolicy`
2. construct `JointMultiAgentTrainer`
3. loop: `collect_rollout` -> `update(rollout, |feats| Some(lambda * redundancy_penalty(feats)))`
4. save checkpoints

`redundancy_penalty` itself (cross-correlation Frobenius norm) can either stay in the example (it's experiment-specific) or move to a public `multi_agent::regularizers` module. Recommend keeping it in the example for now to avoid premature generalization - the trainer only needs to accept any user-supplied `aux_fn`.

## Implementation Plan

1. **Lift types from `train_p3.rs` into `src/multi_agent/joint.rs`**:
   - `Rollout` -> `JointRollout` (rename, make public, add docs)
   - `Config` -> `JointTrainerConfig` (drop experiment-specific fields like `scenario`, `lambda_red`, `output_dir`)
   - `JointStats` -> `JointStats` (expand to include clip_fraction / approx_kl / explained_var per agent, mirroring `TrainingStats` in `src/train/ppo/stats.rs`)
2. **Define the `JointPolicy` trait** and blanket-impl it for `MlpPolicy` and `MultiDiscreteMlpPolicy`. The trait carves out exactly the surface the trainer needs (already present on both policy types today).
3. **Generalize `collect_rollout`**:
   - The example assumes "shared global observation" (Bucket Brigade); the library version should accept distinct per-agent observations from `MultiAgentEnvironment::get_agent_observation(i)`. Provide a convenience constructor for the shared-obs case.
   - Action shape: support both scalar discrete and multi-discrete (`[T]` vs `[T, num_dims]`) - drive off the policy's action tensor shape.
4. **Generalize `ppo_update` -> `update`**:
   - Reuse `compute_policy_loss` / `compute_value_loss` / `generate_minibatch_indices` from `src/train/ppo/loss.rs`.
   - Hook the per-minibatch advantage normalization (the example normalizes once outside the epoch loop; `PPOTrainer` does it inside `train_step_with_aux`. Pick one and document why - probably "outside" is fine for joint since all agents share minibatch indices).
   - Optionally surface per-agent early-stopping on KL like `PPOTrainer` does (it bails when `approx_kl > target_kl`). For joint training, "early-stop when *any* agent breaches" is a reasonable starting policy; document the choice.
5. **Re-export from `src/multi_agent/mod.rs`**:
   ```rust
   pub mod joint;
   pub use joint::{JointMultiAgentTrainer, JointTrainerConfig, JointRollout, JointStats, JointPolicy};
   ```
6. **Migrate `examples/games/bucket_brigade/train_p3.rs`** to use the new trainer. The example should still compile under `--features "training env-bucket-brigade"`. Verify the rewritten example produces numerically equivalent stats on a short run (same seed, same lambda_red, same scenario) - this is the regression bar.
7. **Documentation**:
   - Module docstring on `multi_agent::joint` distinguishing it from `multi_agent::{PolicyLearner, Population}` (synchronized vs independent learners; when to use which).
   - Top-level `src/multi_agent/mod.rs` doc should mention both architectures.

## Acceptance Criteria

- [ ] `src/multi_agent/joint.rs` exists with `JointMultiAgentTrainer`, `JointTrainerConfig`, `JointRollout`, `JointStats`, `JointPolicy` (or equivalent names; bikeshedding welcome).
- [ ] `JointPolicy` blanket-impls land for `MlpPolicy` and `MultiDiscreteMlpPolicy` and `cargo check --features training` succeeds with both used through the trainer.
- [ ] `examples/games/bucket_brigade/train_p3.rs` is refactored to use the library trainer; its `main` is < ~120 lines and the inline `collect_rollout` / `ppo_update` / `Rollout` / `JointStats` definitions are removed (only `redundancy_penalty` plus arg parsing remain experiment-local).
- [ ] Same seed, same scenario, same `lambda_red`, same `num_iterations`: the migrated example produces team-reward and redundancy-loss values that match the pre-migration example within numerical tolerance (e.g. 1e-5 on the first iteration's `total_loss` and `redundancy_loss`).
- [ ] Public API has rustdoc explaining the synchronized-vs-independent contrast with `PolicyLearner` and the single-`backward()` semantics.
- [ ] The trainer accepts an `aux_fn` returning `None` (pure joint IPPO with no cross-agent term) and runs to completion.

## Test Plan

**Unit tests** (`#[cfg(test)] mod tests` inside `src/multi_agent/joint.rs`, gated on the `training` feature):

1. **`test_joint_trainer_smoke`** - Build 2 tiny `MlpPolicy` instances on a mock `MultiAgentEnvironment` (returns deterministic 4-dim obs, scalar rewards), run `collect_rollout(num_steps=64)` and `update(rollout, |_| None)`. Assert: no panics, `JointStats.total_loss.is_finite()`, every per-agent stat is finite, optimizer step count incremented as expected (`n_epochs * num_minibatches`).
2. **`test_joint_rollout_shapes`** - After `collect_rollout` with `T=32`, `N=3`, assert `observations.size() == [32, obs_dim]`, `actions.len() == 3`, `actions[0].size()[0] == 32`, `dones.size() == [32]`.
3. **`test_aux_fn_gradient_couples_encoders`** - Use 2 policies with deterministic init; build a simple `aux_fn` that returns `(features[0] - features[1]).pow(2).sum()`. After one `update`, assert that both policies' encoder weights changed (`var_store` parameter L2 diff > 0 for both). This guards the "shared graph, multiple optimizers" invariant.
4. **`test_aux_fn_none_runs_clean`** - `aux_fn` returns `None`; trainer should not panic and `JointStats.aux_loss == 0.0`.
5. **`test_jointpolicy_for_multidiscrete`** - Same as #1 but with `MultiDiscreteMlpPolicy` and `action_dims=[3, 2]`; verifies the trait works for both policy types.

**Integration / example regression**:

6. **Manual numeric parity check** (documented in PR description, not in CI): run `examples/games/bucket_brigade/train_p3.rs` pre- and post-migration with `--seed 42 --num-iterations 1 --rollout-steps 256 --lambda-red 0.01`; first-iteration `total_loss` and `redundancy_loss` must match to within 1e-5. (CI cannot run libtorch examples on macOS today per issue #8.)

**Existing tests must continue to pass**:

7. `cargo test --features training --lib multi_agent::` (current `multi_agent` module).
8. `cargo test --features training --lib train::ppo::` (PPO loss helpers reused by the joint trainer).
9. `cargo check --example train_p3 --features "training env-bucket-brigade"`.

## Relevant Code References

- **Reference implementation (to be lifted)**: `examples/games/bucket_brigade/train_p3.rs:113-403` - `Rollout`, `collect_rollout`, `compute_gae_single_agent`, `redundancy_penalty`, `ppo_update`, `JointStats`.
- **Existing per-thread architecture (to coexist with, not replace)**: `src/multi_agent/mod.rs`, `src/multi_agent/learner.rs`, `src/multi_agent/population.rs`.
- **PPO building blocks to reuse**: `src/train/ppo/loss.rs::{compute_policy_loss, compute_value_loss, compute_gae, generate_minibatch_indices}`, `src/train/ppo/trainer.rs::PPOTrainer::train_step_with_aux` (single-agent analogue), `src/train/ppo/stats.rs::TrainingStats`.
- **Policy traits/types**: `src/policy/mlp.rs::MlpPolicy::encoder_features`, `src/policy/multi_discrete_mlp.rs::MultiDiscreteMlpPolicy::{get_action, evaluate_actions, encoder_features, optimizer, var_store, device}`.
- **Multi-agent env trait**: `src/multi_agent/environment.rs::MultiAgentEnvironment` (note: currently typed for `actions: &[i64]` - may need extension for multi-discrete actions; coordinate with issue #3).

## Risks / Open Questions

- **`MultiAgentEnvironment::step_multi(&[i64])` is scalar-action only**. The example bypasses this by calling `env.step(&[[u8; 2]])` directly on the concrete `BucketBrigadeMaEnv`. Either (a) extend `MultiAgentEnvironment` to a generic action type / `&[Vec<i64>]`, or (b) ship the joint trainer parameterized over `<E: MultiAgentEnvironment>` AND a separate `<E: MultiDiscreteMultiAgentEnvironment>`. Recommendation: do (a) as a small follow-up if not already covered by issue #3.
- **Shared vs per-agent observation**: the example assumes all agents see the same global view. Most cooperative MARL setups do, but some (Dec-POMDP) don't. The trait already supports `get_agent_observation(i)`; the trainer should call that and store an `[N, T, obs_dim]` tensor, not `[T, obs_dim]`. The shared-obs case is just a special case where all rows are equal. Decide before locking the rollout shape.
- **Tch graph lifetime**: confirmed in the example - holding multiple `Optimizer` instances over the same var-stores and stepping them all after one `backward()` works because each optimizer reads only its own parameters' `.grad`. Document this clearly so readers don't think they need to fork the graph.
- **MPS / device handling**: example pins everything to `policies[0].device()`. Library version should validate that all policies live on the same device and surface a clear error otherwise.

## Related

- Builds on #1 (encoder tap), #2 (aux loss hook for single-agent PPO), #3 (multi-discrete actions), #6 (BB env adapter).
- Reference example: commit `40ec676` introduced `examples/games/bucket_brigade/train_p3.rs`, which is the de facto spec for this trainer.
- Blocked at runtime by #8 (libtorch dylib on macOS) for local validation; the library code itself does not depend on resolving #8 to land.
